src/bson/ext.rs

```diff
@@ -0,0 +1,250 @@
+            Bson::Timestamp(i) => i.to_string(),
+            Bson::Binary(i) => i.to_string(),
+            Bson::ObjectId(i) => i.to_string(),
+            Bson::DateTime(i) => i.to_string(),
+            Bson::Symbol(i) => i.to_string(),
+            Bson::Decimal128(i) => i.to_string(),
+            Bson::Undefined => "undefined".to_string(),
+            _ => return Err(ValueAccessError::UnexpectedType),
+        })
+    }
+
+    fn as_u8(&self) -> Result<u8, ValueAccessError> {
+        Ok(match self {
+            Bson::Double(i) => *i as u8,
+            Bson::Boolean(i) => *i as u8,
+            Bson::Int32(i) => *i as u8,
+            Bson::Int64(i) => *i as u8,
+            _ => return Err(ValueAccessError::UnexpectedType),
+        })
+    }
+
+    fn as_u16(&self) -> Result<u16, ValueAccessError> {
+        Ok(match self {
+            Bson::Double(i) => *i as u16,
+            Bson::Boolean(i) => *i as u16,
+            Bson::Int32(i) => *i as u16,
+            Bson::Int64(i) => *i as u16,
+            _ => return Err(ValueAccessError::UnexpectedType),
+        })
+    }
+
+    fn as_u32(&self) -> Result<u32, ValueAccessError> {
+        Ok(match self {
+            Bson::Double(i) => *i as u32,
+            Bson::Boolean(i) => *i as u32,
+            Bson::Int32(i) => *i as u32,
+            Bson::Int64(i) => *i as u32,
+            Bson::Timestamp(i) => i.time,
+            _ => return Err(ValueAccessError::UnexpectedType),
+        })
+    }
+
+    fn as_u64(&self) -> Result<u64, ValueAccessError> {
+        Ok(match self {
+            Bson::Double(i) => *i as u64,
+            Bson::Boolean(i) => *i as u64,
+            Bson::Int32(i) => *i as u64,
+            Bson::Int64(i) => *i as u64,
+            Bson::Timestamp(i) => i.time as u64,
+            _ => return Err(ValueAccessError::UnexpectedType),
+        })
+    }
+
+    fn to_bytes(self) -> Result<Vec<u8>, ValueAccessError> {
+        Ok(match self {
+            Bson::Binary(i) => i.bytes,
+            _ => return Err(ValueAccessError::UnexpectedType),
+        })
+    }
+
+    fn to_array(self) -> Result<Vec<Bson>, ValueAccessError> {
+        match self {
+            Bson::Array(i) => Ok(i),
+            _ => Err(ValueAccessError::UnexpectedType),
+        }
+    }
+
+    fn to_document(self) -> Result<Document, ValueAccessError> {
+        match self {
+            Bson::Document(i) => Ok(i),
+            _ => Err(ValueAccessError::UnexpectedType),
+        }
+    }
+}
+
+#[derive(Error, Debug)]
+#[allow(missing_docs)]
+pub enum DocError {
+    #[error("Missing key {0}")]
+    MissingKey(String),
+    #[error("Value for key {0} is null")]
+    NullValue(String),
+    #[error(transparent)]
+    ValueAccess(#[from] ValueAccessError),
+    #[error(transparent)]
+    Convert(#[from] mongodb::bson::de::Error),
+}
+
+#[allow(missing_docs)]
+pub trait DocPath {
+    fn split(self) -> Vec<String>;
+}
+
+impl DocPath for &str {
+    fn split(self) -> Vec<String> {
```

Comment:
Maybe `parts` or `path_parts` could be a better name as `split` is already used for something more general.